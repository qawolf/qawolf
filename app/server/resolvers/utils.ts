/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Transaction } from "knex";

import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findGroup } from "../models/group";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { Team, Test, User } from "../types";

type EnsureGroupAccess = {
  group_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureSuiteAccess = {
  logger: Logger;
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  logger: Logger;
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureGroupAccess = async ({
  group_id,
  logger,
  teams,
  trx,
}: EnsureGroupAccess): Promise<Team> => {
  const log = logger.prefix("ensureGroupAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access group", group_id);

  const group = await findGroup(group_id, { logger, trx });
  const selectedTeam = teams!.find((team) => group.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access group", group_id);
    throw new AuthenticationError("cannot access group");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async ({
  logger,
  teams,
  suite_id,
}: EnsureSuiteAccess): Promise<void> => {
  const log = logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, { logger });

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async ({
  logger,
  teams,
  test,
  test_id,
}: EnsureTestAccess): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
