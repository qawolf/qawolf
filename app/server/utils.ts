import { randomBytes, scrypt } from "crypto";
import buildCuid from "cuid";
import * as EmailValidator from "email-validator";
import { v4 as uuidv4 } from "uuid";

import { ClientError } from "./errors";
import { Logger } from "./Logger";

type IsCorrectCode = {
  code: string;
  digest: string;
};

export type RunnerUrl = {
  id: string;
  location: string;
};

export const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
export const API_KEY_PREFIX = "qawolf_";

export const randomChoice = <T>(items: T[] | string): T | string => {
  const randomIndex = Math.floor(Math.random() * items.length);

  return items[randomIndex];
};

export const buildApiKey = (): string => {
  // remove dashes
  const uuid = uuidv4().split("-").join("");

  return `${API_KEY_PREFIX}${uuid}`;
};

// https://dev.to/farnabaz/hash-your-passwords-with-scrypt-using-nodejs-crypto-module-316k
export const buildDigest = async (code: string): Promise<string> => {
  return new Promise((resolve, reject) => {
    const salt = randomBytes(16).toString("hex");

    scrypt(code, salt, 64, (error, derivedKey) => {
      if (error) reject(error);
      resolve(salt + ":" + derivedKey.toString("hex"));
    });
  });
};

export const buildLoginCode = (length = 6): string => {
  let code = "";

  for (let i = 0; i < length; i++) {
    code += randomChoice(ALPHABET);
  }

  return code;
};

export const buildRunnerHost = (options: RunnerUrl): string => {
  if (options.location === "local") return `localhost`;

  return `${options.location}.qawolf.com/runner/${options.id}`;
};

export const buildRunnerStatusUrl = (options: RunnerUrl): string => {
  const protocol = options.location === "local" ? "http" : "https";
  return `${protocol}://${buildRunnerHost(options)}/.qawolf/runner/status`;
};

export const buildRunnerWsUrl = (options: RunnerUrl): string => {
  const protocol = options.location === "local" ? "ws" : "wss";
  return `${protocol}://${buildRunnerHost(options)}/.qawolf`;
};

export const cuid = (): string => buildCuid();

export const isCorrectCode = async ({
  code,
  digest,
}: IsCorrectCode): Promise<boolean> => {
  return new Promise((resolve, reject) => {
    const [salt, key] = digest.split(":");

    scrypt(code, salt, 64, (error, derivedKey) => {
      if (error) reject(error);
      resolve(key === derivedKey.toString("hex"));
    });
  });
};

// https://stackoverflow.com/a/1214753
export const minutesFromNow = (minutes = 0): string => {
  return new Date(Date.now() + minutes * 60000).toISOString();
};

export const validateEmail = (email: string, logger: Logger): void => {
  const isValid = EmailValidator.validate(email);
  if (isValid) return;

  logger.error(`email ${email} invalid`);
  throw new ClientError("invalid email address");
};
