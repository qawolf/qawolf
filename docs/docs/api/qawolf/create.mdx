---
id: create
sidebar_label: create
title: qawolf.create
---

import ArgumentCreate from '../components/qawolf/ArgumentCreate';
import ReturnVoid from '../components/qawolf/ReturnVoid';
import Syntax from '../components/Syntax';

<Syntax code="create([options])" />

Convert your browser actions to code. Call `create` in a test or script file to insert new code where it was called.

See the [guide on adding steps to existing tests](../../add_steps_to_existing_tests) to learn more about using `create`.

### Arguments

<ArgumentCreate />

### Returns

<ReturnVoid />

### Examples

Create new code in an existing file, after calling [`qawolf.register(context)`](register) to register a [Playwright `BrowserContext`](https://github.com/microsoft/playwright/blob/master/docs/api.md#class-browsercontext) instance:

```js
const qawolf = require('qawolf');
// ...
// somewhere earlier in file
qawolf.register(context);

await qawolf.create();
```

Create new code in an existing file, passing a [Playwright `BrowserContext`](https://github.com/microsoft/playwright/blob/master/docs/api.md#class-browsercontext) instance directly:

```js
const qawolf = require('qawolf');
// ...
await qawolf.create({ context });
```

Create new code in an existing file, passing a [Playwright `BrowserContext`](https://github.com/microsoft/playwright/blob/master/docs/api.md#class-browsercontext) instance directly. Use a different selector path than the default:

```js
const qawolf = require('qawolf');
// ...
await qawolf.create({ context, selectorPath: 'my/path/selectors.json' });
```
