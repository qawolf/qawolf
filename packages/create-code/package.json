{
  "name": "@qawolf/create-code",
  "description": "qawolf: create code",
  "version": "0.9.0",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@qawolf/build-code": "file:../build-code",
    "@qawolf/build-workflow": "file:../build-workflow",
    "@qawolf/logger": "file:../logger",
    "@qawolf/repl": "file:../repl"
  },
  "devDependencies": {
    "@qawolf/test": "file:../test",
    "@qawolf/types": "file:../types"
  },
  "peerDependencies": {
    "fs-extra": "^8.x",
    "kleur": "^3.x",
    "lodash": "^4.x"
  }
}
