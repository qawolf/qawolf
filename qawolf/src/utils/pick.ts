// Since we only need this for now, we can avoid downloading lodash
const pick = (
  obj: Record<any, any>,
  props: readonly string[]
): Record<any, any> => {
  const newObj = {};
  for (const prop of props) {
    if ({}.hasOwnProperty.call(obj, prop)) {
      newObj[prop] = obj[prop];
    }
  }
  return newObj;
};

export default pick;
