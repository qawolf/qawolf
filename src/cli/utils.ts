import { parse, UrlWithStringQuery } from 'url';

export const omitArgs = (args: string[], argsToOmit: string[]): string[] => {
  return args.filter(arg => !argsToOmit.some(skip => arg.startsWith(skip)));
};

export const parseUrl = (urlString: string): UrlWithStringQuery => {
  let url = parse(urlString);

  // prefix w/ https if a protocol is not provided
  if (!url.protocol) {
    url = parse(`https://${urlString}`);
  }

  if (!url.hostname) {
    throw new Error(`Invalid url ${urlString}`);
  }

  return url;
};
