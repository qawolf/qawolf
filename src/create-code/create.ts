import callsites from 'callsites';
import Debug from 'debug';
import { pathExists, readFile } from 'fs-extra';
import { bold } from 'kleur';
import { findLast } from 'lodash';
import { basename, dirname, join } from 'path';
import { BrowserContext } from 'playwright';
import { ReplContext } from 'playwright-utils';
import { CREATE_HANDLE } from './CodeUpdater';
import { CreateManager } from './CreateManager';
import { getLineIncludes } from './format';

type CreateOptions = {
  codePath?: string;
  context?: BrowserContext;
  // used for testing
  onReady?: () => void;
  selectorPath?: string;
};

const debug = Debug('qawolf:create');

export const getCodePath = async (
  callerFileNames: string[],
): Promise<string> => {
  debug(`search caller files for ${CREATE_HANDLE} %j`, callerFileNames);

  const codes = await Promise.all(
    callerFileNames.map(async filename => {
      let code = '';

      if (await pathExists(filename)) {
        code = await readFile(filename, 'utf8');
      }

      return { code, filename };
    }),
  );

  const item = findLast(
    codes,
    ({ code }) => !!getLineIncludes(code, CREATE_HANDLE),
  );

  if (!item) {
    throw new Error(`Could not find ${CREATE_HANDLE} in caller`);
  }

  return item.filename;
};

export const getSelectorPath = (codePath: string): string => {
  const codeName = basename(codePath).split('.')[0];
  return join(dirname(codePath), '../selectors', `${codeName}.json`);
};

export const create = async (options: CreateOptions = {}): Promise<void> => {
  const context: BrowserContext =
    options.context || (ReplContext.data() as any).context;
  if (!context) {
    throw new Error(
      'No context found. Call qawolf.register(context) before qawolf.create() or qawolf.create({ context })',
    );
  }

  let codePath = options.codePath;
  if (!codePath) {
    const callerFileNames = callsites().map(c => c.getFileName());
    codePath = await getCodePath(callerFileNames);
  }

  const selectorPath = options.selectorPath || getSelectorPath(codePath);

  const manager = await CreateManager.create({
    codePath,
    context,
    selectorPath,
  });

  console.log(bold().blue('üê∫  QA Wolf is ready to create code!'));
  if (options.onReady) options.onReady();

  await manager.finalize();
};
