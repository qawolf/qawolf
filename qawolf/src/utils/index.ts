import pick from "./pick";

export { pick };
