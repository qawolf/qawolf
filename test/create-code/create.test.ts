import { readFile } from 'fs-extra';
import { prompt } from 'inquirer';
import { join, relative } from 'path';
import { launch, register, waitFor } from 'playwright-utils';
import { getSelectorPath, getCodePath } from '../../src/create-code/create';
import { createSelf, getCallSites } from '../e2e/scripts/createSelf';
import { TEST_URL } from '../utils';

jest.mock('inquirer');

describe('create', () => {
  it('works end-to-end', async () => {
    let fulfillPrompt: ({ choice: string }) => void;
    ((prompt as unknown) as jest.Mock).mockResolvedValue(
      new Promise<{ choice: string }>(resolve => (fulfillPrompt = resolve)),
    );

    const filePath = join(__dirname, '../e2e/scripts/createSelf.ts');
    const loadCode = (): Promise<string> => readFile(filePath, 'utf8');

    const initial = await loadCode();
    // check the initial code has not changed
    expect(initial).toMatchSnapshot('1-initial');

    // set up context
    const browser = await launch();
    const context = await browser.newContext();
    await register(context);
    const page = await context.newPage();
    await page.goto(`${TEST_URL}text-inputs`);

    // wait until the creation is ready
    let createdPromise: Promise<void>;
    await new Promise(onReady => {
      createdPromise = createSelf(onReady);
    });

    // check the patch handle replaces the create handle
    expect(await loadCode()).toMatchSnapshot('2-prepared');

    // perform an action
    await page.type("[data-qa='html-text-input']", 'hello');

    // check the code updated as expected
    await waitFor(async () => (await loadCode()).includes('hello'));
    expect(await loadCode()).toMatchSnapshot('3-updated');

    // discard the changes
    fulfillPrompt({ choice: 'Discard' });
    await createdPromise;

    const reverted = await loadCode();
    expect(reverted).toEqual(initial);
    expect(reverted).toMatchSnapshot('4-reverted');

    await browser.close();
  });
});

describe('getCodePath', () => {
  it('finds the caller file with create handle', async () => {
    const codePath = await getCodePath(getCallSites());
    const relativePath = relative(__filename, codePath);
    expect(relativePath).toEqual('../../e2e/scripts/createSelf.ts');
  });
});

describe('getSelectorPath', () => {
  it('returns a sibling path for the code file', () => {
    const selectorPath = getSelectorPath(__filename);
    const relativePath = relative(__filename, selectorPath);
    expect(relativePath).toEqual('../../selectors/create.json');
  });
});
