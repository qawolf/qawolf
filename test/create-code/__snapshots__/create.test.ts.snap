// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create works end-to-end: 1-initial 1`] = `
"import callsites from 'callsites';
import * as qawolf from '../../../src';

export const createSelf = async (onReady: () => void): Promise<void> => {
  // 🐺 I code, therefore I am
  await qawolf.create({ onReady });
};

export const getCallSites = (): string[] =>
  callsites().map(c => c.getFileName());
"
`;

exports[`create works end-to-end: 2-prepared 1`] = `
"import callsites from 'callsites';
import * as qawolf from '../../../src';

export const createSelf = async (onReady: () => void): Promise<void> => {
  // 🐺 I code, therefore I am
  // 🐺 CREATE CODE HERE
};

export const getCallSites = (): string[] =>
  callsites().map(c => c.getFileName());
"
`;

exports[`create works end-to-end: 3-updated 1`] = `
"import callsites from 'callsites';
import * as qawolf from '../../../src';

export const createSelf = async (onReady: () => void): Promise<void> => {
  // 🐺 I code, therefore I am
  await page.type(\\"[data-qa='html-text-input']\\", \\"hello\\");
  // 🐺 CREATE CODE HERE
};

export const getCallSites = (): string[] =>
  callsites().map(c => c.getFileName());
"
`;

exports[`create works end-to-end: 4-reverted 1`] = `
"import callsites from 'callsites';
import * as qawolf from '../../../src';

export const createSelf = async (onReady: () => void): Promise<void> => {
  // 🐺 I code, therefore I am
  await qawolf.create({ onReady });
};

export const getCallSites = (): string[] =>
  callsites().map(c => c.getFileName());
"
`;
